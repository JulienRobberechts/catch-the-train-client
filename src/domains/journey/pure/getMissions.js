import stationsNANI from "../../../data/ratp/rers/A/stations-NANI.json";
import stationsNOTE from "../../../data/ratp/rers/A/stations-NOTE.json";
import stationsQIKI from "../../../data/ratp/rers/A/stations-QIKI.json";

export default function getMissions(departureStation, arrivalStation) {
  const missionsSchedules = getMissionsSchedules();
  const missions = calculateMissionsForJourney(
    missionsSchedules,
    departureStation,
    arrivalStation
  );
  return missions;
}

export function getMissionsSchedules() {
  const missionsSchedules = [
    { mission: "NANI", ...stationsNANI },
    { mission: "NOTE", ...stationsNOTE },
    { mission: "QIKI", ...stationsQIKI },
  ];
  const formattedSchedule = missionsSchedules.map(({ mission, result }) => ({
    mission,
    stations: result.stations.map((station) => station.slug),
  }));
  return formattedSchedule;
}

export function calculateMissionsForJourney(
  missionsSchedules,
  departureStation,
  arrivalStation
) {
  const missions = missionsSchedules
    .map(addStationIndexToSchedule(departureStation, "departureStationIndex"))
    .filter((schedule) => schedule.departureStationIndex >= 0)
    .map(addStationIndexToSchedule(arrivalStation, "arrivalStationIndex"))
    .filter((schedule) => schedule.arrivalStationIndex >= 0)
    .filter(
      (schedule) =>
        schedule.arrivalStationIndex >= schedule.departureStationIndex
    )
    .map((schedule) => schedule.mission);

  return missions;
}

const addStationIndexToSchedule = (station, indexPropName) => ({
  stations,
  ...otherProps
}) => ({
  [indexPropName]: stations.findIndex((s) => s === station),
  stations,
  ...otherProps,
});

export { addStationIndexToSchedule };
