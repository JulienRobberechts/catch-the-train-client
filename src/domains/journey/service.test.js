import each from "jest-each";
import getMissions, { getMissionsSchedules } from "./service";

describe("journey services", () => {
  describe("getMissionsSchedules", () => {
    test("should get Missions Schedules", () => {
      const missionsSchedules = getMissionsSchedules();
      expect(missionsSchedules).toMatchSnapshot();
    });
  });

  describe("getMissions", () => {
    each([
      ["cergy+prefecture", "houilles", ["NANI"]],
      ["chatelet+les+halles", "sucy+bonneuil", ["NANI", "NOTE"]],
      ["maisons+laffitte", "nation", ["NANI", "NOTE"]],
      ["auber", "gare+de+lyon", ["NANI", "NOTE", "QIKI"]],
      ["charles+de+gaulle+etoile", "marne+la+vallee+chessy", ["QIKI"]],
    ]).test(
      "should get missions from '%s' to '%s'",
      (departure, arrival, expectedMissions) => {
        const actualMissions = getMissions(departure, arrival);
        expect(actualMissions).toEqual(expectedMissions);
      }
    );
  });
});
