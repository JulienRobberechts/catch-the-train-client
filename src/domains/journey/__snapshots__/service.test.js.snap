// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`journey services getMissionsSchedules should get Missions Schedules 1`] = `
Array [
  Object {
    "mission": "TAPI",
    "stations": Array [
      "boissy+saint+leger",
      "sucy+bonneuil",
      "la+varenne+chennevieres",
      "champigny",
      "le+parc+de+saint+maur",
      "saint+maur+creteil",
      "joinville+le+pont",
      "nogent+sur+marne",
      "fontenay+sous+bois",
      "vincennes",
      "nation",
      "gare+de+lyon",
      "chatelet+les+halles",
      "auber",
      "charles+de+gaulle+etoile",
      "grande+arche+la+defense",
      "nanterre+prefecture",
      "houilles",
      "sartrouville",
      "maisons+laffitte",
      "acheres+grand+cormier",
      "poissy",
    ],
  },
  Object {
    "mission": "NANI",
    "stations": Array [
      "cergy+le+haut",
      "cergy+st+christophe",
      "cergy+prefecture",
      "neuville+universite",
      "conflans+fin+d'oise",
      "acheres+ville",
      "maisons+laffitte",
      "sartrouville",
      "houilles",
      "nanterre+prefecture",
      "grande+arche+la+defense",
      "charles+de+gaulle+etoile",
      "auber",
      "chatelet+les+halles",
      "gare+de+lyon",
      "nation",
      "vincennes",
      "fontenay+sous+bois",
      "nogent+sur+marne",
      "joinville+le+pont",
      "saint+maur+creteil",
      "le+parc+de+saint+maur",
      "champigny",
      "la+varenne+chennevieres",
      "sucy+bonneuil",
      "boissy+saint+leger",
    ],
  },
  Object {
    "mission": "NATO",
    "stations": Array [
      "cergy+le+haut",
      "cergy+st+christophe",
      "cergy+prefecture",
      "neuville+universite",
      "conflans+fin+d'oise",
      "acheres+ville",
      "maisons+laffitte",
      "sartrouville",
      "houilles",
      "nanterre+prefecture",
      "grande+arche+la+defense",
      "charles+de+gaulle+etoile",
      "auber",
      "chatelet+les+halles",
      "gare+de+lyon",
      "nation",
      "vincennes",
      "fontenay+sous+bois",
      "nogent+sur+marne",
      "joinville+le+pont",
      "saint+maur+creteil",
      "le+parc+de+saint+maur",
      "champigny",
      "la+varenne+chennevieres",
      "sucy+bonneuil",
      "boissy+saint+leger",
    ],
  },
  Object {
    "mission": "NELY",
    "stations": Array [
      "st+germain+en+laye",
      "le+vesinet+le+pecq",
      "le+vesinet+centre",
      "chatou+croissy",
      "rueil+malmaison",
      "nanterre+ville",
      "nanterre+universite",
      "nanterre+prefecture",
      "grande+arche+la+defense",
      "charles+de+gaulle+etoile",
      "auber",
      "chatelet+les+halles",
      "gare+de+lyon",
      "nation",
      "vincennes",
      "fontenay+sous+bois",
      "nogent+sur+marne",
      "joinville+le+pont",
      "saint+maur+creteil",
      "le+parc+de+saint+maur",
      "champigny",
      "la+varenne+chennevieres",
      "sucy+bonneuil",
      "boissy+saint+leger",
    ],
  },
  Object {
    "mission": "NEMO",
    "stations": Array [
      "st+germain+en+laye",
      "le+vesinet+le+pecq",
      "le+vesinet+centre",
      "chatou+croissy",
      "rueil+malmaison",
      "nanterre+ville",
      "nanterre+universite",
      "nanterre+prefecture",
      "grande+arche+la+defense",
      "charles+de+gaulle+etoile",
      "auber",
      "chatelet+les+halles",
      "gare+de+lyon",
      "nation",
      "vincennes",
      "fontenay+sous+bois",
      "nogent+sur+marne",
      "joinville+le+pont",
      "saint+maur+creteil",
      "le+parc+de+saint+maur",
      "champigny",
      "la+varenne+chennevieres",
      "sucy+bonneuil",
      "boissy+saint+leger",
    ],
  },
  Object {
    "mission": "NOTE",
    "stations": Array [
      "poissy",
      "acheres+grand+cormier",
      "maisons+laffitte",
      "sartrouville",
      "houilles",
      "nanterre+prefecture",
      "grande+arche+la+defense",
      "charles+de+gaulle+etoile",
      "auber",
      "chatelet+les+halles",
      "gare+de+lyon",
      "nation",
      "vincennes",
      "fontenay+sous+bois",
      "nogent+sur+marne",
      "joinville+le+pont",
      "saint+maur+creteil",
      "le+parc+de+saint+maur",
      "champigny",
      "la+varenne+chennevieres",
      "sucy+bonneuil",
      "boissy+saint+leger",
    ],
  },
  Object {
    "mission": "QHAR",
    "stations": Array [
      "rueil+malmaison",
      "nanterre+ville",
      "nanterre+universite",
      "nanterre+prefecture",
      "grande+arche+la+defense",
      "charles+de+gaulle+etoile",
      "auber",
      "chatelet+les+halles",
      "gare+de+lyon",
      "nation",
      "vincennes",
      "val+de+fontenay",
      "neuilly+plaisance",
      "bry+sur+marne",
      "noisy+le+grand+mont+d'est",
      "noisy+champs",
      "noisiel",
      "lognes",
      "torcy",
      "bussy+st+georges",
      "val+d'europe+serris",
      "marne+la+vallee+chessy",
    ],
  },
  Object {
    "mission": "QIKI",
    "stations": Array [
      "st+germain+en+laye",
      "le+vesinet+le+pecq",
      "le+vesinet+centre",
      "chatou+croissy",
      "rueil+malmaison",
      "nanterre+ville",
      "nanterre+universite",
      "nanterre+prefecture",
      "grande+arche+la+defense",
      "charles+de+gaulle+etoile",
      "auber",
      "chatelet+les+halles",
      "gare+de+lyon",
      "nation",
      "vincennes",
      "val+de+fontenay",
      "neuilly+plaisance",
      "bry+sur+marne",
      "noisy+le+grand+mont+d'est",
      "noisy+champs",
      "noisiel",
      "lognes",
      "torcy",
      "bussy+st+georges",
      "val+d'europe+serris",
      "marne+la+vallee+chessy",
    ],
  },
  Object {
    "mission": "QIWI",
    "stations": Array [
      "st+germain+en+laye",
      "le+vesinet+le+pecq",
      "le+vesinet+centre",
      "chatou+croissy",
      "rueil+malmaison",
      "nanterre+ville",
      "nanterre+universite",
      "nanterre+prefecture",
      "grande+arche+la+defense",
      "charles+de+gaulle+etoile",
      "auber",
      "chatelet+les+halles",
      "gare+de+lyon",
      "nation",
      "vincennes",
      "val+de+fontenay",
      "neuilly+plaisance",
      "bry+sur+marne",
      "noisy+le+grand+mont+d'est",
      "noisy+champs",
      "noisiel",
      "lognes",
      "torcy",
      "bussy+st+georges",
      "val+d'europe+serris",
      "marne+la+vallee+chessy",
    ],
  },
  Object {
    "mission": "QUDO",
    "stations": Array [
      "poissy",
      "acheres+grand+cormier",
      "maisons+laffitte",
      "sartrouville",
      "houilles",
      "nanterre+prefecture",
      "grande+arche+la+defense",
      "charles+de+gaulle+etoile",
      "auber",
      "chatelet+les+halles",
      "gare+de+lyon",
      "nation",
      "vincennes",
      "val+de+fontenay",
      "neuilly+plaisance",
      "bry+sur+marne",
      "noisy+le+grand+mont+d'est",
      "noisy+champs",
      "noisiel",
      "lognes",
      "torcy",
      "bussy+st+georges",
      "val+d'europe+serris",
      "marne+la+vallee+chessy",
    ],
  },
  Object {
    "mission": "QURE",
    "stations": Array [
      "poissy",
      "acheres+grand+cormier",
      "maisons+laffitte",
      "sartrouville",
      "houilles",
      "nanterre+prefecture",
      "grande+arche+la+defense",
      "charles+de+gaulle+etoile",
      "auber",
      "chatelet+les+halles",
      "gare+de+lyon",
      "nation",
      "vincennes",
      "val+de+fontenay",
      "neuilly+plaisance",
      "bry+sur+marne",
      "noisy+le+grand+mont+d'est",
      "noisy+champs",
      "noisiel",
      "lognes",
      "torcy",
      "bussy+st+georges",
      "val+d'europe+serris",
      "marne+la+vallee+chessy",
    ],
  },
  Object {
    "mission": "QYAN",
    "stations": Array [
      "cergy+le+haut",
      "cergy+st+christophe",
      "cergy+prefecture",
      "neuville+universite",
      "conflans+fin+d'oise",
      "acheres+ville",
      "maisons+laffitte",
      "sartrouville",
      "houilles",
      "nanterre+prefecture",
      "grande+arche+la+defense",
      "charles+de+gaulle+etoile",
      "auber",
      "chatelet+les+halles",
      "gare+de+lyon",
      "nation",
      "vincennes",
      "val+de+fontenay",
      "neuilly+plaisance",
      "bry+sur+marne",
      "noisy+le+grand+mont+d'est",
      "noisy+champs",
      "noisiel",
      "lognes",
      "torcy",
      "bussy+st+georges",
      "val+d'europe+serris",
      "marne+la+vallee+chessy",
    ],
  },
  Object {
    "mission": "QYLT",
    "stations": Array [
      "cergy+le+haut",
      "cergy+st+christophe",
      "cergy+prefecture",
      "neuville+universite",
      "conflans+fin+d'oise",
      "acheres+ville",
      "maisons+laffitte",
      "sartrouville",
      "houilles",
      "nanterre+prefecture",
      "grande+arche+la+defense",
      "charles+de+gaulle+etoile",
      "auber",
      "chatelet+les+halles",
      "gare+de+lyon",
      "nation",
      "vincennes",
      "val+de+fontenay",
      "neuilly+plaisance",
      "bry+sur+marne",
      "noisy+le+grand+mont+d'est",
      "noisy+champs",
      "noisiel",
      "lognes",
      "torcy",
      "bussy+st+georges",
      "val+d'europe+serris",
      "marne+la+vallee+chessy",
    ],
  },
  Object {
    "mission": "TAXE",
    "stations": Array [
      "boissy+saint+leger",
      "sucy+bonneuil",
      "la+varenne+chennevieres",
      "champigny",
      "le+parc+de+saint+maur",
      "saint+maur+creteil",
      "joinville+le+pont",
      "nogent+sur+marne",
      "fontenay+sous+bois",
      "vincennes",
      "nation",
      "gare+de+lyon",
      "chatelet+les+halles",
      "auber",
      "charles+de+gaulle+etoile",
      "grande+arche+la+defense",
      "nanterre+prefecture",
      "houilles",
      "sartrouville",
      "maisons+laffitte",
      "acheres+grand+cormier",
      "poissy",
    ],
  },
  Object {
    "mission": "TEDI",
    "stations": Array [
      "marne+la+vallee+chessy",
      "val+d'europe+serris",
      "bussy+st+georges",
      "torcy",
      "lognes",
      "noisiel",
      "noisy+champs",
      "noisy+le+grand+mont+d'est",
      "bry+sur+marne",
      "neuilly+plaisance",
      "val+de+fontenay",
      "vincennes",
      "nation",
      "gare+de+lyon",
      "chatelet+les+halles",
      "auber",
      "charles+de+gaulle+etoile",
      "grande+arche+la+defense",
      "nanterre+prefecture",
      "houilles",
      "sartrouville",
      "maisons+laffitte",
      "acheres+grand+cormier",
      "poissy",
    ],
  },
  Object {
    "mission": "TERI",
    "stations": Array [
      "marne+la+vallee+chessy",
      "val+d'europe+serris",
      "bussy+st+georges",
      "torcy",
      "lognes",
      "noisiel",
      "noisy+champs",
      "noisy+le+grand+mont+d'est",
      "bry+sur+marne",
      "neuilly+plaisance",
      "val+de+fontenay",
      "vincennes",
      "nation",
      "gare+de+lyon",
      "chatelet+les+halles",
      "auber",
      "charles+de+gaulle+etoile",
      "grande+arche+la+defense",
      "nanterre+prefecture",
      "houilles",
      "sartrouville",
      "maisons+laffitte",
      "acheres+grand+cormier",
      "poissy",
    ],
  },
  Object {
    "mission": "TETE",
    "stations": Array [
      "marne+la+vallee+chessy",
      "val+d'europe+serris",
      "bussy+st+georges",
      "torcy",
      "lognes",
      "noisiel",
      "noisy+champs",
      "noisy+le+grand+mont+d'est",
      "bry+sur+marne",
      "neuilly+plaisance",
      "val+de+fontenay",
      "vincennes",
      "nation",
      "gare+de+lyon",
      "chatelet+les+halles",
      "auber",
      "charles+de+gaulle+etoile",
      "grande+arche+la+defense",
      "nanterre+prefecture",
      "houilles",
      "sartrouville",
      "maisons+laffitte",
      "acheres+grand+cormier",
      "poissy",
    ],
  },
  Object {
    "mission": "UPAC",
    "stations": Array [
      "marne+la+vallee+chessy",
      "val+d'europe+serris",
      "bussy+st+georges",
      "torcy",
      "lognes",
      "noisiel",
      "noisy+champs",
      "noisy+le+grand+mont+d'est",
      "bry+sur+marne",
      "neuilly+plaisance",
      "val+de+fontenay",
      "vincennes",
      "nation",
      "gare+de+lyon",
      "chatelet+les+halles",
      "auber",
      "charles+de+gaulle+etoile",
      "grande+arche+la+defense",
      "nanterre+prefecture",
      "houilles",
      "sartrouville",
      "maisons+laffitte",
      "acheres+ville",
      "conflans+fin+d'oise",
      "neuville+universite",
      "cergy+prefecture",
      "cergy+st+christophe",
      "cergy+le+haut",
    ],
  },
  Object {
    "mission": "UPUS",
    "stations": Array [
      "marne+la+vallee+chessy",
      "val+d'europe+serris",
      "bussy+st+georges",
      "torcy",
      "lognes",
      "noisiel",
      "noisy+champs",
      "noisy+le+grand+mont+d'est",
      "bry+sur+marne",
      "neuilly+plaisance",
      "val+de+fontenay",
      "vincennes",
      "nation",
      "gare+de+lyon",
      "chatelet+les+halles",
      "auber",
      "charles+de+gaulle+etoile",
      "grande+arche+la+defense",
      "nanterre+prefecture",
      "houilles",
      "sartrouville",
      "maisons+laffitte",
      "acheres+ville",
      "conflans+fin+d'oise",
      "neuville+universite",
      "cergy+prefecture",
      "cergy+st+christophe",
      "cergy+le+haut",
    ],
  },
  Object {
    "mission": "UZAR",
    "stations": Array [
      "boissy+saint+leger",
      "sucy+bonneuil",
      "la+varenne+chennevieres",
      "champigny",
      "le+parc+de+saint+maur",
      "saint+maur+creteil",
      "joinville+le+pont",
      "nogent+sur+marne",
      "fontenay+sous+bois",
      "vincennes",
      "nation",
      "gare+de+lyon",
      "chatelet+les+halles",
      "auber",
      "charles+de+gaulle+etoile",
      "grande+arche+la+defense",
      "nanterre+prefecture",
      "houilles",
      "sartrouville",
      "maisons+laffitte",
      "acheres+ville",
      "conflans+fin+d'oise",
      "neuville+universite",
      "cergy+prefecture",
      "cergy+st+christophe",
      "cergy+le+haut",
    ],
  },
  Object {
    "mission": "UZEL",
    "stations": Array [
      "boissy+saint+leger",
      "sucy+bonneuil",
      "la+varenne+chennevieres",
      "champigny",
      "le+parc+de+saint+maur",
      "saint+maur+creteil",
      "joinville+le+pont",
      "nogent+sur+marne",
      "fontenay+sous+bois",
      "vincennes",
      "nation",
      "gare+de+lyon",
      "chatelet+les+halles",
      "auber",
      "charles+de+gaulle+etoile",
      "grande+arche+la+defense",
      "nanterre+prefecture",
      "houilles",
      "sartrouville",
      "maisons+laffitte",
      "acheres+ville",
      "conflans+fin+d'oise",
      "neuville+universite",
      "cergy+prefecture",
      "cergy+st+christophe",
      "cergy+le+haut",
    ],
  },
  Object {
    "mission": "ZEBU",
    "stations": Array [
      "boissy+saint+leger",
      "sucy+bonneuil",
      "la+varenne+chennevieres",
      "champigny",
      "le+parc+de+saint+maur",
      "saint+maur+creteil",
      "joinville+le+pont",
      "nogent+sur+marne",
      "fontenay+sous+bois",
      "vincennes",
      "nation",
      "gare+de+lyon",
      "chatelet+les+halles",
      "auber",
      "charles+de+gaulle+etoile",
      "grande+arche+la+defense",
      "nanterre+prefecture",
      "nanterre+universite",
      "nanterre+ville",
      "rueil+malmaison",
      "chatou+croissy",
      "le+vesinet+centre",
      "le+vesinet+le+pecq",
      "st+germain+en+laye",
    ],
  },
  Object {
    "mission": "ZEMA",
    "stations": Array [
      "boissy+saint+leger",
      "sucy+bonneuil",
      "la+varenne+chennevieres",
      "champigny",
      "le+parc+de+saint+maur",
      "saint+maur+creteil",
      "joinville+le+pont",
      "nogent+sur+marne",
      "fontenay+sous+bois",
      "vincennes",
      "nation",
      "gare+de+lyon",
      "chatelet+les+halles",
      "auber",
      "charles+de+gaulle+etoile",
      "grande+arche+la+defense",
      "nanterre+prefecture",
      "nanterre+universite",
      "nanterre+ville",
      "rueil+malmaison",
      "chatou+croissy",
      "le+vesinet+centre",
      "le+vesinet+le+pecq",
      "st+germain+en+laye",
    ],
  },
  Object {
    "mission": "ZEUS",
    "stations": Array [
      "boissy+saint+leger",
      "sucy+bonneuil",
      "la+varenne+chennevieres",
      "champigny",
      "le+parc+de+saint+maur",
      "saint+maur+creteil",
      "joinville+le+pont",
      "nogent+sur+marne",
      "fontenay+sous+bois",
      "vincennes",
      "nation",
      "gare+de+lyon",
      "chatelet+les+halles",
      "auber",
      "charles+de+gaulle+etoile",
      "grande+arche+la+defense",
      "nanterre+prefecture",
      "nanterre+universite",
      "nanterre+ville",
      "rueil+malmaison",
      "chatou+croissy",
      "le+vesinet+centre",
      "le+vesinet+le+pecq",
      "st+germain+en+laye",
    ],
  },
  Object {
    "mission": "ZULI",
    "stations": Array [
      "marne+la+vallee+chessy",
      "val+d'europe+serris",
      "bussy+st+georges",
      "torcy",
      "lognes",
      "noisiel",
      "noisy+champs",
      "noisy+le+grand+mont+d'est",
      "bry+sur+marne",
      "neuilly+plaisance",
      "val+de+fontenay",
      "vincennes",
      "nation",
      "gare+de+lyon",
      "chatelet+les+halles",
      "auber",
      "charles+de+gaulle+etoile",
      "grande+arche+la+defense",
      "nanterre+prefecture",
      "nanterre+universite",
      "nanterre+ville",
      "rueil+malmaison",
      "chatou+croissy",
      "le+vesinet+centre",
      "le+vesinet+le+pecq",
      "st+germain+en+laye",
    ],
  },
]
`;
